core/src/main/scala/org/apache/spark/scheduler/HealthTracker.scala

```diff
@@ -0,0 +1,478 @@
+          s"since allocation client is not defined.")
+    }
+  }
+
+  private def killExcludedExecutor(exec: String): Unit = {
+    if (conf.get(config.EXCLUDE_ON_FAILURE_KILL_ENABLED)) {
+      killExecutor(exec, s"Killing excluded executor id $exec since " +
+        s"${config.EXCLUDE_ON_FAILURE_KILL_ENABLED.key} is set.")
+    }
+  }
+
+  private[scheduler] def killExcludedIdleExecutor(exec: String): Unit = {
+    killExecutor(exec,
+      s"Killing excluded idle executor id $exec because of task unschedulability and trying " +
+        "to acquire a new executor.")
+  }
+
+  private def killExecutorsOnExcludedNode(node: String): Unit = {
+    if (conf.get(config.EXCLUDE_ON_FAILURE_KILL_ENABLED)) {
+      allocationClient match {
+        case Some(a) =>
+          logInfo(s"Killing all executors on excluded host $node " +
+            s"since ${config.EXCLUDE_ON_FAILURE_KILL_ENABLED.key} is set.")
+          if (a.killExecutorsOnHost(node) == false) {
+            logError(s"Killing executors on node $node failed.")
+          }
+        case None =>
+          logWarning(s"Not attempting to kill executors on excluded host $node " +
+            s"since allocation client is not defined.")
+      }
+    }
+  }
+
+  def updateExcludedForFetchFailure(host: String, exec: String): Unit = {
+    if (EXCLUDE_FETCH_FAILURE_ENABLED) {
+      // If we exclude on fetch failures, we are implicitly saying that we believe the failure is
+      // non-transient, and can't be recovered from (even if this is the first fetch failure,
+      // stage is retried after just one failure, so we don't always get a chance to collect
+      // multiple fetch failures).
+      // If the external shuffle-service is on, then every other executor on this node would
+      // be suffering from the same issue, so we should exclude (and potentially kill) all
+      // of them immediately.
+
+      val now = clock.getTimeMillis()
+      val expiryTimeForNewExcludes = now + EXCLUDE_ON_FAILURE_TIMEOUT_MILLIS
+
+      if (conf.get(config.SHUFFLE_SERVICE_ENABLED)) {
+        if (!nodeIdToExcludedExpiryTime.contains(host)) {
+          logInfo(s"excluding node $host due to fetch failure of external shuffle service")
+
+          nodeIdToExcludedExpiryTime.put(host, expiryTimeForNewExcludes)
+          listenerBus.post(SparkListenerNodeExcluded(now, host, 1))
+          _excludedNodeList.set(nodeIdToExcludedExpiryTime.keySet.toSet)
+          killExecutorsOnExcludedNode(host)
+          updateNextExpiryTime()
+        }
+      } else if (!executorIdToExcludedStatus.contains(exec)) {
+        logInfo(s"Excluding executor $exec due to fetch failure")
+
+        executorIdToExcludedStatus.put(exec, ExcludedExecutor(host, expiryTimeForNewExcludes))
+        // We hardcoded number of failure tasks to 1 for fetch failure, because there's no
+        // reattempt for such failure.
+        listenerBus.post(SparkListenerExecutorExcluded(now, exec, 1))
```

Comment:
depends on what is looking for it. Nothing in spark is looking for it anymore. If others are listening for it and don't change then yes. This is a developer api though.  I kind of hated to have to post 2 separate messages which is why I removed it.
Do you know of people actively relying on those listeners?

The only thing internal to Spark that it breaks is that older history servers don't see blacklisted on the new history files.

Reply:
Can we provide a default listener implementation that delegates the event to the old method?

Reply:
Yeah, I'm talking about custom listeners. I don't exactly know how people rely on it but I guess it won't be used a lot.

 Maybe, we can just remove it for simplicity as `SparkListener` (developer API) and `SparkListenerInterface` ( private spark) don't guarantee the backward compatibility?

Reply:
so I left it in the listeners for 2 reasons. 
1)  SparkFirehoseListener is public, though I think that is mostly for testing purposes. Maybe this being public means SparkListenerExecutorExcluded is public so we should just make sure its sent.
2) for backwards compatibility of the history server reading the old event logs containing those events.

So really the only thing I got rid of is what you pointed out here with us posting it on new applications. I'm fine either way if we think its going to break a bunch of people we can post 2 messages. I think I would rather do that then have the default listener delegate just in case something weird happens there.

Reply:
> 1. SparkFirehoseListener is public...

I feel like this a historical problem... The only way to add a custom listener is by using `sc.addSparkListener(listener: SparkListenerInterface)`. And `SparkListenerInterface` is decorated by `private[spark]`. So, it's impossible to keep backward compatibility of `SparkFirehoseListener` when `SparkListenerInterface` changes APIs. Therefore, it's meaningless to say that `SparkFirehoseListener` is public without considering `SparkListenerInterface`. 
BTW, `SparkFirehoseListener` is actually a Java version of `SparkListener`. I think it should be developer API too but somehow we miss it.

> 2. for backwards compatibility of the history server reading the old event logs containing those events.

This's a good point. Then, I think we can remove the interface `onExecutorBlacklisted(executorBlacklisted: SparkListenerExecutorBlacklisted)` but can't remove the event `SparkListenerExecutorBlacklisted`. And for history server, I think we can override `onOtherEvent` to handle old `SparkListenerExecutorBlacklisted` specifically in order to keep backward compatibility.

> So really the only thing I got rid of is what you pointed out here with us posting it on new applications. I'm fine either way if we think its going to break a bunch of people we can post 2 messages. I think I would rather do that then have the default listener delegate just in case something weird happens there.

My original point is that it's meaningless to keep those APIs while we don't post `SparkListenerExecutorBlacklisted` anymore. We should either remove all things completely or still post old events with those APIs kept. I'm fine either way but just not stay in the middle. Though, I personally prefer to remove those APIs as mentioned above(Spark has no backward compatibility guarantee for the listener).

Reply:
sounds good, I'll move the handling to OnOtherEvent

Reply:
* Both `SparkListener` and `SparkFirehoseListener` are exposed classes : extending `SparkListenerInterface` (we missed adding `DeveloperApi` to `SparkFirehoseListener`).
  * `sc.addSparkListener()` should work fine for these user supplied subclasses.
  * We will have subclasses depending on these callbacks in user codebases - including consuming existing blacklist events this way.
* We can currently make forward compatible changes in `SparkListenerInterface`.

The issue we have here is a backwardly incompatible change - this is an issue for the exposed api.
While we can make backwardly incompatible changes to `SparkListenerInterface` directly (like dropping callbacks), the exposed subclasses should continue to preserve behavior while we go through deprecation cycle.
Until we remove support, this will mean we also rewire the new events to call the old handler methods until we drop support in 4.0 (?).

Reply:
cc @gatorsmile